### ISSUE TITLE
Questions: Multiple instances

### ISSUE THREAD

**Issue by jacodv** [jamesmh/coravel#101]
**Global Scheduled tasks**
This might be a broader question to just using Coravel as a task scheduler.  How would schedule tasks that need to run once a for the application, but the service is actually running on more than one node in a load-balanced environment.  

**Possible solution**
When the first node gets that `Invoke` it sets a record in the database that the task was started.  Subsequent tasks that get invoked todo the same daily job then stops after determining that another process is handling the task.

**Potential issue**
* Race condition if all tasks fire the exact same moment
* Retry / re-invoke if the task did not complete successfully and need to be picked up by another node/service

**Comment by jamesmh** [comment]
This is a feature I have planned to add - a distributed scheduler.

Another option right now is, if you know how many instances you have (how many nodes), then configure an env var for each one and put an if statement inside your scheduling logic. Only run certain tasks on certain nodes (on one node each).

Another, if possible, is to create a background service (console app) for all your schedules.

But yes, this would be a great feature that I've been hoping to build 👍

**Comment by jacodv** [comment]
@jamesmh: Wow great to get a fast response.  I did some more research and it seems that a database/caching mechanism is the best option.  I am going to insert a "Started" document into a MongoDB with a key that will be consistent across all instances, this will allow me to have only one instance that can insert the document (who will do the work), and the rest failing with a duplicate key error.

From your documentation I suggest, we add a CanExecute(<Action<bool>>) allowing the implementation to return a boolean if this instance can start the process (task).

We do many things that your library makes easier like queues (RabbitMQ), Caching (MongoDB) etc.  I will keep this project in mind.

**Comment by jamesmh** [comment]
Nice. I was thinking of taking the approach that laravel takes (which I've already pre-architected out in Coravel's code). Right now, for example, calling `PreventOverlap` internally uses a shared mutex (by interface) to make these checks.

For a distributed scheduler, I need to create some distributed mutexes (one per concrete DB provider). The difficulty will be in maintaining all the various providers...so my struggle is to start with the most commonly used technology (SQL Server, Postgres, Mongo, Redis?)

My first hunch is that Postgres would be, overall, the best first step. The mechanics would be very close to what you had mentioned - using a table lock pattern where each row represents a lock for a specific task/job.

My second hunch is to start with Redis since it's a pretty "neutral" ground and a fairly easy tech to get up and running from the developer's perspective. If someone needs a distributed scheduler then chances are they already recognize that there's a commitment to a little extra infrastructure that will be needed.

**Comment by jacodv** [comment]
Sounds like a solid approach!  👍

**Comment by jacodv** [comment]
@jamesmh: I decided to implement our cross servers task using your package as a base.  I have a strange issue that is baffling me.  I cannot get the configuration settings in the web startup.  I need this to configure the scheduled tasks.  Have you encountered this before?  Or have a suggestion using your solution to solve the issue?

See the sample project as reference:  
[SettingInStartUpTest.zip](https://github.com/jamesmh/coravel/files/3456307/SettingInStartUpTest.zip)

**Comment by jamesmh** [comment]
Awesome. Haven't had a chance to check it out. Glad you nailed it down!

**Comment by caizhiyuan** [comment]
When will this feature be available?

**Comment by jamesmh** [comment]
Probably not anytime soon TBH 😥 I'd like to but not a priority among life's various challenges right now 😉

**Comment by Memhave** [comment]
I'm also interesting in this - does any body know how it handles itself on a cluster like kubernetes with multiple replicas?

**Comment by jamesmh** [comment]
Each instance will execute the schedules. So if 5 instances exist (for example), then 5 jobs will be executed in parallel.

**Comment by johnwc** [comment]
Any ETA on when this feature can be introduced? It is what is keeping us from moving from hangire to Coravel.
